ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ranger/RangerRestClientImpl.java

```diff
@@ -0,0 +1,356 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.core.MediaType;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.File;
+import java.io.Reader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ArrayList;
+
+/**
+ * RangerRestClientImpl to connect to Ranger and export policies.
+ */
+public class RangerRestClientImpl implements RangerRestClient {
+  private static final Logger LOG = LoggerFactory.getLogger(RangerRestClientImpl.class);
+  private static final String RANGER_REST_URL_EXPORTJSONFILE = "/service/plugins/policies/exportJson";
+  private static final String RANGER_REST_URL_IMPORTJSONFILE =
+      "/service/plugins/policies/importPoliciesFromFile?updateIfExists=true";
+
+  public RangerExportPolicyList exportRangerPolicies(String sourceRangerEndpoint,
+                                                     String dbName, String rangerHiveServiceName) throws Exception {
+    LOG.info("Ranger endpoint for cluster " + sourceRangerEndpoint);
+    ClientResponse clientResp = null;
+    String uri = null;
+    if (!StringUtils.isEmpty(rangerHiveServiceName)) {
+      uri = RANGER_REST_URL_EXPORTJSONFILE + "?serviceName=" + rangerHiveServiceName + "&polResource="
+          + dbName + "&resource:database=" + dbName
+          + "&serviceType=hive&resourceMatchScope=self_or_ancestor&resourceMatch=full";
+    }
+    if (sourceRangerEndpoint.endsWith("/")) {
+      sourceRangerEndpoint = StringUtils.removePattern(sourceRangerEndpoint, "/+$");
+    }
+    String url = sourceRangerEndpoint + (uri != null ? (uri.startsWith("/") ? uri : ("/" + uri)) : "");
+    LOG.debug("URL to export policies from source Ranger: {}", url);
+    RangerExportPolicyList rangerExportPolicyList = new RangerExportPolicyList();
+    WebResource.Builder builder = getRangerResourceBuilder(url);
+    clientResp = builder.get(ClientResponse.class);
+
+    String response = null;
+    if (clientResp != null) {
+      if (clientResp.getStatus() == HttpServletResponse.SC_OK) {
+        Gson gson = new GsonBuilder().create();
+        response = clientResp.getEntity(String.class);
+        LOG.debug("Response received for ranger export {} ", response);
+        if (StringUtils.isNotEmpty(response)) {
+          rangerExportPolicyList = gson.fromJson(response, RangerExportPolicyList.class);
+          return rangerExportPolicyList;
+        }
+      } else if (clientResp.getStatus() == HttpServletResponse.SC_NO_CONTENT) {
+        LOG.debug("Ranger policy export request returned empty list");
+        return rangerExportPolicyList;
+      } else if (clientResp.getStatus() == HttpServletResponse.SC_UNAUTHORIZED) {
```

Comment:
Should we introduce new exception MetadatReplException or so, and use that during metadata replication?

Reply:
Using SemanticException